File: tkpdinbox/src/main/java/com/tokopedia/inbox/inboxchat/viewholder/AttachedInvoiceSelectionViewHolder.java
package com.tokopedia.inbox.inboxchat.viewholder;

import android.support.annotation.LayoutRes;
import android.support.v7.widget.RecyclerView;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.tokopedia.abstraction.common.utils.image.ImageHandler;
import com.tokopedia.core.base.adapter.viewholders.AbstractViewHolder;
import com.tokopedia.core.var.TkpdState;
import com.tokopedia.inbox.R;
import com.tokopedia.inbox.attachinvoice.view.viewholder.InvoiceViewHolder;
import com.tokopedia.inbox.inboxchat.listener.AttachedInvoiceOnSelectedListener;
import com.tokopedia.inbox.inboxchat.presenter.ChatRoomContract;
import com.tokopedia.inbox.inboxchat.viewmodel.AttachInvoiceSelectionViewModel;
import com.tokopedia.inbox.inboxchat.viewmodel.AttachInvoiceSentViewModel;
import com.tokopedia.inbox.inboxchat.viewmodel.AttachInvoiceSingleViewModel;
import com.tokopedia.inbox.inboxchat.viewmodel.mapper.AttachInvoiceMapper;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by Hendri on 28/03/18.
 */

public class AttachedInvoiceSelectionViewHolder  extends AbstractViewHolder<AttachInvoiceSelectionViewModel> {

    @LayoutRes
    public static final int LAYOUT = 2;
    private ChatRoomContract.View selectedListener;
    private AttachedInvoiceSelectionViewHolder.AttachedInvoicesItemsAdapter singleItemAdapter;

    public AttachedInvoiceSelectionViewHolder(View itemView,ChatRoomContract.View selectedListener) {
        super(itemView);
        this.selectedListener = selectedListener;
        singleItemAdapter = new AttachedInvoicesItemsAdapter();
    }

    @Override
    public void bind(AttachInvoiceSelectionViewModel element) {
        singleItemAdapter.setList(element.getList());
    }

    private class AttachedInvoicesItemsAdapter extends RecyclerView.Adapter<AttachedInvoiceSingleItemViewHolder>{
        List<AttachInvoiceSingleViewModel> list;
        @Override
        public AttachedInvoiceSingleItemViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
            View itemView = LayoutInflater.from(parent.getContext())
                    .inflate(R.layout.item_carousel_invoice_attach, parent, false);
            return new AttachedInvoiceSingleItemViewHolder(itemView);
        }

        @Override
        public void onBindViewHolder(AttachedInvoiceSingleItemViewHolder holder, int position) {
            holder.bind(list.get(position));
            holder.itemView.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    selectedListener.onInvoiceSelected(
                            AttachInvoiceMapper.selectedInvoiceViewModelToSelectedInvoice(list.get(getAdapterPosition()))
                    );
                }
            });
        }

        @Override
        public int getItemCount() {
            return list.size();
        }

        public List<AttachInvoiceSingleViewModel> getList() {
            return list;
        }

        public void setList(List<AttachInvoiceSingleViewModel> list) {
            this.list = list;
            notifyDataSetChanged();
        }
    }

    private class AttachedInvoiceSingleItemViewHolder extends RecyclerView.ViewHolder{

        private TextView invoiceNo;
        private TextView invoiceDate;
        private TextView productName;
        private TextView productDesc;
        private TextView invoiceStatus;
        private TextView totalAmount;
        private ImageView productImage;

        public AttachedInvoiceSingleItemViewHolder(View itemView) {
            super(itemView);
            invoiceNo = itemView.findViewById(R.id.attach_invoice_item_invoice_no);
            invoiceDate = itemView.findViewById(R.id.attach_invoice_item_invoice_date);
            productName = itemView.findViewById(R.id.attach_invoice_item_product_name);
            productDesc = itemView.findViewById(R.id.attach_invoice_item_product_desc);
            invoiceStatus = itemView.findViewById(R.id.attach_invoice_item_invoice_status);
            totalAmount = itemView.findViewById(R.id.attach_invoice_item_invoice_total);
            productImage = itemView.findViewById(R.id.attach_invoice_item_product_image);
        }

        public void bind(AttachInvoiceSingleViewModel element){
            invoiceNo.setText(element.getCode());
            if(!TextUtils.isEmpty(element.getImageUrl())) {
                productImage.setVisibility(View.VISIBLE);
                ImageHandler.loadImageAndCache(productImage, element.getImageUrl());
            }
            else {
                productImage.setVisibility(View.GONE);
            }
            invoiceDate.setText(element.getCreatedTime());
            productName.setText(element.getTitle());
            productDesc.setText(element.getDescription());
            invoiceStatus.setText(element.getStatus());
            totalAmount.setText(element.getAmount());
        }
    }
}

File: features/user/chatbot/src/main/java/com/tokopedia/chatbot/view/adapter/viewholder/AttachedInvoiceSelectionViewHolder.kt
package com.tokopedia.chatbot.view.adapter.viewholder

import android.text.TextUtils
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.annotation.LayoutRes
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.tokopedia.abstraction.base.view.adapter.viewholders.AbstractViewHolder
import com.tokopedia.abstraction.common.utils.image.ImageHandler
import com.tokopedia.chatbot.ChatbotConstant.RENDER_INVOICE_LIST_AND_BUTTON_ACTION
import com.tokopedia.chatbot.R
import com.tokopedia.chatbot.attachinvoice.domain.mapper.AttachInvoiceMapper
import com.tokopedia.chatbot.data.invoice.AttachInvoiceSelectionUiModel
import com.tokopedia.chatbot.data.invoice.AttachInvoiceSingleUiModel
import com.tokopedia.chatbot.databinding.ItemChatbotCarouselInvoiceAttachBinding
import com.tokopedia.chatbot.databinding.ItemChatbotInvoiceAttachSelectionBinding
import com.tokopedia.chatbot.view.adapter.viewholder.listener.AttachedInvoiceSelectionListener
import com.tokopedia.chatbot.view.util.InvoiceStatusLabelHelper
import com.tokopedia.kotlin.extensions.view.hide
import com.tokopedia.kotlin.extensions.view.invisible
import com.tokopedia.kotlin.extensions.view.show
import com.tokopedia.unifycomponents.ImageUnify
import com.tokopedia.unifycomponents.Label
import com.tokopedia.unifycomponents.UnifyButton

/**
 * Created by Hendri on 28/03/18.
 */

class AttachedInvoiceSelectionViewHolder(
    itemView: View,
    private val selectedListener: AttachedInvoiceSelectionListener
) :
    AbstractViewHolder<AttachInvoiceSelectionUiModel>(itemView) {
    private val view = ItemChatbotInvoiceAttachSelectionBinding.bind(itemView)
    private val singleItemAdapter: AttachedInvoiceSelectionViewHolder.AttachedInvoicesItemsAdapter
    private val invoiceSelection: RecyclerView = view.attachInvoiceChatInvoiceSelection

    init {
        singleItemAdapter = AttachedInvoicesItemsAdapter()
        invoiceSelection.layoutManager = LinearLayoutManager(
            itemView.context,
            LinearLayoutManager.HORIZONTAL,
            false
        )
        invoiceSelection.adapter = singleItemAdapter
    }

    override fun bind(element: AttachInvoiceSelectionUiModel) {
        if (element.status == RENDER_INVOICE_LIST_AND_BUTTON_ACTION) {
            singleItemAdapter.setList(element.list)
            invoiceSelection.show()
        } else {
            invoiceSelection.hide()
        }
    }

    private inner class AttachedInvoicesItemsAdapter : RecyclerView.Adapter<AttachedInvoiceSingleItemViewHolder>() {
        internal var list: List<AttachInvoiceSingleUiModel>? = null

        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): AttachedInvoiceSingleItemViewHolder {
            val itemView = ItemChatbotCarouselInvoiceAttachBinding.inflate(LayoutInflater.from(parent.context), parent, false)
            return AttachedInvoiceSingleItemViewHolder(itemView)
        }

        override fun onBindViewHolder(
            holder: AttachedInvoiceSingleItemViewHolder,
            position: Int
        ) {
            list?.getOrNull(position)?.let { holder.bind(it) }
            holder.pilihButton.setOnClickListener {
                selectedListener.onInvoiceSelected(
                    AttachInvoiceMapper.invoiceViewModelToDomainInvoicePojo(list!![position])
                )
            }
        }

        override fun getItemCount(): Int {
            return list?.size ?: 0
        }

        fun getList(): List<AttachInvoiceSingleUiModel>? {
            return list
        }

        fun setList(list: List<AttachInvoiceSingleUiModel>?) {
            this.list = list
            notifyDataSetChanged()
        }
    }

    private inner class AttachedInvoiceSingleItemViewHolder internal constructor(itemView: ItemChatbotCarouselInvoiceAttachBinding) : RecyclerView.ViewHolder(itemView.root) {

        private val invoiceDate: TextView
        private val productName: TextView
        private val productDesc: TextView
        private val invoiceStatus: Label
        private val price: TextView
        private val productImage: ImageUnify
        private val pricePrefix: TextView
        val pilihButton: UnifyButton

        init {
            invoiceDate = itemView.containerAllInvoiceAttach.tvInvoiceDate
            productName = itemView.containerAllInvoiceAttach.tvInvoiceName
            productDesc = itemView.containerAllInvoiceAttach.tvInvoiceDesc
            invoiceStatus = itemView.containerAllInvoiceAttach.tvStatus
            pricePrefix = itemView.containerAllInvoiceAttach.tvPricePrefix
            price = itemView.containerAllInvoiceAttach.tvPrice
            productImage = itemView.containerAllInvoiceAttach.ivThumbnail
            pilihButton = itemView.containerAllInvoiceAttach.btnPilih
        }

        fun bind(element: AttachInvoiceSingleUiModel) {
            if (!TextUtils.isEmpty(element.imageUrl)) {
                productImage.show()
                ImageHandler.loadImageRounded2(productImage.context, productImage, element.imageUrl)
            } else {
                productImage.hide()
            }
            invoiceDate.text = element.createdTime
            productName.text = element.title
            setProductDesc(element.description)
            setStatus(element.status, element.color, element.statusId)
            setPrice(element.amount)
        }

        private fun setProductDesc(description: String) {
            if (description.isNotEmpty()) {
                productDesc.show()
                productDesc.text = description
            } else {
                productDesc.hide()
            }
        }

        private fun setPrice(totalAmount: String?) {
            if (totalAmount.isNullOrEmpty()) {
                pricePrefix.hide()
                price.hide()
            } else {
                pricePrefix.show()
                price.text = totalAmount
                price.show()
            }
        }

        private fun setStatus(status: String, statusColor: String?, statusId: Int) {
            if (status.isNotEmpty() == true) {
                var labelType: Int = if (statusColor != null && statusColor.isEmpty()) {
                    InvoiceStatusLabelHelper.getLabelTypeWithStatusId(statusId)
                } else {
                    InvoiceStatusLabelHelper.getLabelType(statusColor)
                }
                invoiceStatus.text = status
                invoiceStatus.setLabelType(labelType)
                invoiceStatus.show()
            } else {
                invoiceStatus.invisible()
            }
        }
    }

    companion object {
        private val ADDITIONAL_GET_ALL_BUTTON = 1

        @LayoutRes
        val LAYOUT = R.layout.item_chatbot_invoice_attach_selection
    }
}
